plugins/comment/comment.7
.TH YED-COMMENT 7 "YED Plugin Manuals" "" "YED Plugin Manuals"
.SH NAME
comment \- Toggles commenting of lines of code.
.SH CONFIGURATION
.SS cpp-style-comments [ yes / no ]
If this variable is true, single line C++ style comments will be used.
.SH COMMANDS
.SS comment-toggle
Toggles the current line. If there is a selection, toggles every line in the selection.
.SH BUFFERS
None
.SH NOTES
Implemented filetypes are: C, C++, Config, Golang, Zig, Shell, bJou, Python, yedrc, LaTeX
.SH VERSION
0.0.1
.SH KEYWORDS
toggle, comment


CHANGELOG.md
# Changelog

## 1306 - 2021-12-13
### Fixed
    - fixed a bug that would cause command line contents to overflow into a new terminal line.

## 1305 - 2021-12-13
### Fixed
    - fixed a bug that would cause some cursor movements to crash on binary buffer contents.

## 1304 - 2021-12-10
### Added
    - new command `frame` activates frame based on their number.

## 1303 - 2021-12-8
### Changed
    - the `plugins` directory is now a `subtree`. This makes installing simpler and allows you to just grab a `tar` of a release without needing `git`.

## 1302 - 2021-12-7
### Changed
    - removed use of `codesign` in the install script for Darwin targets. This isn't needed any more and will help us package `yed`.

## 1301 - 2021-12-7
### Fixed
    - Fixed logic in yed_cell_is_in_frame()

## 1300 - 2021-12-6
### Added
- Started keeping a changelog.
- Started tagging versions in git.
- A new syntax highlighting engine in `syntax.h`.
- Status line additions:
    - `%{...}` syntax for expanding a variable that contains an attribute string
    - Padding and justification on `%` expansions: `%` followed by an integer, `x`, will be padded to `x` characters. If `x` is preceded by a `-`, the text will be left justified. A `=` indicates center justification. The default is right justification. This has no effect on expansions that result in attributes.
    - `%T`: 24-hour clock
- Events:
    - `EVENT_STATUS_LINE_PRE_UPDATE` will trigger before the status line is about to be updated.
    - `EVENT_PLUGIN_MESSAGE` is a new general purpose event that plugins can send and other plugins can handle.
- Mouse support added to the core for left, center, and right buttons as well as up and down scroll wheel. Distinguishes between press, release, and drag.
- Useful functions for getting buffer contents as byte strings:
    - `char *yed_get_selection_text(yed_buffer *buffer);`
    - `char *yed_get_line_text(yed_buffer *buffer, int row);`
    - `char *yed_get_buffer_text(yed_buffer *buffer);`
- Style components:
    - `good`: general indication in the positive (e.g. a notification that a build succeeded)
    - `bad`: general indication in the negative (e.g. a notification that a build failed)
    - `popup`: the foreground and background to be used for popup menus
    - `popup-alt`: the foreground and background to be used for other popup menu elements (e.g. the selected item)
    - `code-variable`: for variables in code
    - `code-field`: for fields of a structure or class
    - `code-escape`: for escape sequences in string literals
### Changed
- `highlight.h` is now deprecated and will be removed in version 1400.
- Status line: `%T`, which used to expand to the filetype, now expands to a 24-hour clock. `%F` now expands to the filetype.
### Removed
- Removed the default command `if`.
### Fixed
- A buffer's undo history is now reset when `buffer-reload` gets run (or whenever the contents of the buffer are loaded from a file).
- Plugins are now install-agnostic -- they will use whatever `libyed.so` is being used by the binary. This fixes a crash that would occur when installations would mix on the same system.


comment/comment.c
#include <yed/plugin.h>

void comment_toggle(int n_args, char **args);
int  comment_toggle_line(yed_frame *frame, yed_line *line, int row);

void comment_toggle_line_c(yed_frame *frame, yed_line *line, int row);
void comment_line_c(yed_frame *frame, int row);
void uncomment_line_c(yed_frame *frame, int row);

void comment_toggle_line_hash(yed_frame *frame, yed_line *line, int row);
void comment_line_hash(yed_frame *frame, int row);
void uncomment_line_hash(yed_frame *frame, int row);

void comment_toggle_line_latex(yed_frame *frame, yed_line *line, int row);
void comment_line_latex(yed_frame *frame, int row);
void uncomment_line_latex(yed_frame *frame, int row);

void comment_toggle_line_cpp(yed_frame *frame, yed_line *line, int row);
void comment_line_cpp(yed_frame *frame, int row);
void uncomment_line_cpp(yed_frame *frame, int row);

int yed_plugin_boot(yed_plugin *self) {
    YED_PLUG_VERSION_CHECK();

    if (yed_get_var("cpp-style-comments") == NULL) {
        yed_set_var("cpp-style-comments", "no");
    }
    yed_plugin_set_command(self, "comment-toggle", comment_toggle);
    return 0;
}

void comment_toggle(int n_args, char **args) {
    yed_frame  *frame;
    yed_buffer *buff;
    yed_line   *line;
    int         save_col,
                row,
                r1, c1, r2, c2,
                status;

    if (n_args != 0) {
        yed_cerr("expected 0 arguments but got %d", n_args);
        return;
    }

    if (!ys->active_frame) {
        yed_cerr("no active frame");
        return;
    }

    frame = ys->active_frame;

    if (!frame->buffer) {
        yed_cerr("active frame has no buffer");
        return;
    }

    buff = frame->buffer;

    yed_start_undo_record(frame, buff);

    save_col = frame->cursor_col;
    yed_set_cursor_within_frame(frame, frame->cursor_line, 1);

    if (buff->has_selection) {
        yed_range_sorted_points(&buff->selection, &r1, &c1, &r2, &c2);

        for (row = r1; row <= r2; row += 1) {
            line   = yed_buff_get_line(buff, row);
            status = comment_toggle_line(frame, line, row);
        }

        frame->dirty = 1;

        YEXE("select-off");
    } else {
        line   = yed_buff_get_line(buff, frame->cursor_line);
        status = comment_toggle_line(frame, line, frame->cursor_line);
    }

    yed_set_cursor_within_frame(frame, frame->cursor_line, save_col);

    if (status == 0) {
        yed_cerr("unhandled ft");
        yed_cancel_undo_record(frame, buff);
    } else {
        yed_end_undo_record(frame, buff);
    }
}

int comment_toggle_line(yed_frame *frame, yed_line *line, int row) {
    if (       frame->buffer->ft == yed_get_ft("C")       ||
               frame->buffer->ft == yed_get_ft("C++")     ||
               frame->buffer->ft == yed_get_ft("Golang")) {
        if (yed_var_is_truthy("cpp-style-comments")){
            comment_toggle_line_cpp(frame, line, row);
        } else {
            comment_toggle_line_c(frame, line, row);
        }

    } else if (frame->buffer->ft == yed_get_ft("Shell")  ||
               frame->buffer->ft == yed_get_ft("bJou")   ||
               frame->buffer->ft == yed_get_ft("Config") ||
               frame->buffer->ft == yed_get_ft("Python") ||
               frame->buffer->ft == yed_get_ft("yedrc")) {
        comment_toggle_line_hash(frame, line, row);

    } else if (frame->buffer->ft == yed_get_ft("LaTeX")) {
        comment_toggle_line_latex(frame, line, row);

    } else if (frame->buffer->ft == yed_get_ft("Zig"))   {
        comment_toggle_line_cpp(frame, line, row);

    } else {
        return 0;
    }

    return 1;
}


/* C */
void comment_toggle_line_c(yed_frame *frame, yed_line *line, int row) {
    int        line_len;
    yed_glyph *g;

    line_len = line->visual_width;

    /* Are we uncommenting? */
    if (line_len >= 6) {
        g = yed_line_col_to_glyph(line, 1);
        if (g->c == '/') { g = yed_line_col_to_glyph(line, 2);
        if (g->c == '*') { g = yed_line_col_to_glyph(line, 3);
        if (g->c == ' ') { g = yed_line_col_to_glyph(line, line_len);
        if (g->c == '/') { g = yed_line_col_to_glyph(line, line_len - 1);
        if (g->c == '*') { g = yed_line_col_to_glyph(line, line_len - 2);
        if (g->c == ' ') {
            uncomment_line_c(frame, row);
            return;
        }}}}}}
    }

    comment_line_c(frame, row);
}

void comment_line_c(yed_frame *frame, int row) {
    yed_insert_into_line(frame->buffer, row, 1, G(' '));
    yed_insert_into_line(frame->buffer, row, 1, G('*'));
    yed_insert_into_line(frame->buffer, row, 1, G('/'));

    yed_append_to_line(frame->buffer, row, G(' '));
    yed_append_to_line(frame->buffer, row, G('*'));
    yed_append_to_line(frame->buffer, row, G('/'));
}

void uncomment_line_c(yed_frame *frame, int row) {
    yed_delete_from_line(frame->buffer, row, 1);
    yed_delete_from_line(frame->buffer, row, 1);
    yed_delete_from_line(frame->buffer, row, 1);

    yed_pop_from_line(frame->buffer, row);
    yed_pop_from_line(frame->buffer, row);
    yed_pop_from_line(frame->buffer, row);
}

/* Hash-comment languages */
void comment_toggle_line_hash(yed_frame *frame, yed_line *line, int row) {
    int        line_len;
    yed_glyph *g;

    line_len = line->visual_width;

    /* Are we uncommenting? */
    if (line_len >= 2) {
        g = yed_line_col_to_glyph(line, 1);
        if (g->c == '#') {
            g = yed_line_col_to_glyph(line, 2);
            if (g->c == ' ') {
                uncomment_line_hash(frame, row);
                return;
            }
        }
    }

    comment_line_hash(frame, row);
}

void comment_line_hash(yed_frame *frame, int row) {
    yed_insert_into_line(frame->buffer, row, 1, G(' '));
    yed_insert_into_line(frame->buffer, row, 1, G('#'));
}

void uncomment_line_hash(yed_frame *frame, int row) {
    yed_delete_from_line(frame->buffer, row, 1);
    yed_delete_from_line(frame->buffer, row, 1);
}

/* LaTeX */
void comment_toggle_line_latex(yed_frame *frame, yed_line *line, int row) {
    int        line_len;
    yed_glyph *g;

    line_len = line->visual_width;

    /* Are we uncommenting? */
    if (line_len >= 2) {
        g = yed_line_col_to_glyph(line, 1);
        if (g->c == '%') {
            g = yed_line_col_to_glyph(line, 2);
            if (g->c == ' ') {
                uncomment_line_hash(frame, row);
                return;
            }
        }
    }

    comment_line_latex(frame, row);
}

void comment_line_latex(yed_frame *frame, int row) {
    yed_insert_into_line(frame->buffer, row, 1, G(' '));
    yed_insert_into_line(frame->buffer, row, 1, G('%'));
}

void uncomment_line_latex(yed_frame *frame, int row) {
    yed_delete_from_line(frame->buffer, row, 1);
    yed_delete_from_line(frame->buffer, row, 1);
}

/* CPP style comments*/
void comment_toggle_line_cpp(yed_frame *frame, yed_line *line, int row) {
    int        line_len;
    yed_glyph *g;

    line_len = line->visual_width;

    /* Are we uncommenting? */
    if (line_len >= 3) {
        g = yed_line_col_to_glyph(line, 1);
        if (g->c == '/') {
            g = yed_line_col_to_glyph(line, 2);
            if (g->c == '/') {
                g = yed_line_col_to_glyph(line, 3);
                if (g->c == ' ') {
                    uncomment_line_cpp(frame, row);
                    return;
                }
            }
        }
    }
    comment_line_cpp(frame, row);
}

void comment_line_cpp(yed_frame *frame, int row) {
    yed_insert_into_line(frame->buffer, row, 1, G(' '));
    yed_insert_into_line(frame->buffer, row, 1, G('/'));
    yed_insert_into_line(frame->buffer, row, 1, G('/'));
}

void uncomment_line_cpp(yed_frame *frame, int row) {
    yed_delete_from_line(frame->buffer, row, 1);
    yed_delete_from_line(frame->buffer, row, 1);
    yed_delete_from_line(frame->buffer, row, 1);
}
